### DISCUSSION TITLE
tokio::join vs tokio::task::JoinSet

### DISCUSSION THREAD

**Discussion by joelchen** [tokio-rs/tokio#6921] [Q&A] [answered]
What are the differences between using the join macro and using JoinSet for running multiple asynchronous tasks concurrently and waiting for them to finish? Do they have the same performance?
```
let t1_handler = tokio::spawn(t1());
let t2_handler = tokio::spawn(t2());
if let (Err(e1), Err(e2)) = tokio::join!(t1_handler, t2_handler) {
    println!("error: {}, {}", e1, e2);
}
```
```
let mut set = tokio::task::JoinSet::new();
set.spawn(t1());
set.spawn(t2());
set.join_all().await;
```
Similar issue found but it is not using the join macro, instead using JoinHandle returned from tokio::spawn pushed into FuturesUnordered: https://github.com/tokio-rs/tokio/issues/5564

**Comment by Darksonn** [comment, marked as answer]
It doesn't make any sense to use `tokio::join!` in that example. It's just a complicated, slower way to write this:
```rs
let t1_handler = tokio::spawn(t1());
let t2_handler = tokio::spawn(t2());
let res1 = t1_handler.await;
let res2 = t2_handler.await;
```
This will be faster than `JoinSet`. The time when you should use `JoinSet` is when you need to act on tasks exiting immediately. If you're just waiting for all of them to exit, there's no real advantage over using `JoinHandle` directly.
